apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Tests/StorageClientImplTest.cs

```diff
@@ -69,5 +76,418 @@ public void GetEffectiveEncryptionKey_Invalid(EncryptionKey clientkey, Encryptio
+
+        // Object test cases BEGIN
+
+        [Fact]
+        public void GetObject_NoRetry() =>
+            NoRetryHelper(service => service.Objects.Get("bucket", "object"), client => client.GetObject("bucket", "object"));
+
+        [Fact]
+        public void GetObject_RetryOnce() =>
+            RetryOnceHelper(service => service.Objects.Get("bucket", "object"), client => client.GetObject("bucket", "object"));
+
+        [Fact]
+        public void GetObject_RetryThenFail() =>
+            RetryThenFailHelper(service => service.Objects.Get("bucket", "object"), client => client.GetObject("bucket", "object"));
+
+
+        /*
+         * /copy object test cases are currently failing
+         * Also not sure where to add  IfGenerationMatch condition to test
+         * Error:
+         Assert.Equal() Failure
+        Expected: https://storage.googleapis.com/storage/v1/b/sourceBucket/o/sourceObject/copyTo/b/destinationBucket/o/destinationObject
+        Actual:   https://storage.googleapis.com/storage/v1/b/sourceBucket/o/sourceObject/rewriteTo/b/destinationBucket/o/destinationObject
+         */
+
+        [Fact]
+        public void CopyObject_NoRetry() =>
+      NoRetryHelper(service => service.Objects.Copy(new Object() { Id = "id", Name = "name" }, "sourceBucket", "sourceObject", "destinationBucket", "destinationObject"), client => client.CopyObject("sourceBucket", "sourceObject", "destinationBucket", "destinationObject"));
+
+        [Fact]
+        public void CopyObject_RetryOnce() =>
+RetryOnceHelper(service => service.Objects.Copy(new Object() { Id = "id", Name = "name" }, "sourceBucket", "sourceObject", "destinationBucket", "destinationObject"), client => client.CopyObject("sourceBucket", "sourceObject", "destinationBucket", "destinationObject"));
+
+        [Fact]
+        public void CopyObject_RetryThenFail() =>
+    RetryThenFailHelper(service => service.Objects.Copy(new Object() { Id = "id", Name = "name" }, "sourceBucket", "sourceObject", "destinationBucket", "destinationObject"), client => client.CopyObject("sourceBucket", "sourceObject", "destinationBucket", "destinationObject"));
+
+        [Fact]
+        public void ListObject_NoRetry() =>
+    NoRetryHelper(service => service.Objects.List("bucket"), client => client.ListObjects("bucket").Count());
+
+        [Fact]
+        public void ListObject_RetryOnce() =>
+    RetryOnceHelper(service => service.Objects.List("bucket"), client => client.ListObjects("bucket").Count());
+
+        [Fact]
+        public void ListObject_RetryThenFail() =>
+    RetryThenFailHelper(service => service.Objects.List("bucket"), client => client.ListObjects("bucket").Count());
+
+        [Fact]
+        public void DeleteObject_NoRetry() =>
+    NoRetryHelper(service => service.Objects.Delete("bucket", "objectName"), client => client.DeleteObject("bucket", "objectName"));
+
+        // Need to create overloaded helper method for below operations to run:
+
+        /*
```

Comment:
Right. I'm suggesting that you have:

```csharp
private static void RetryOnceHelper<T>(Func<StorageService, ClientServiceRequest<T>> requestProvider, Action<StorageClient> clientAction,
   HttpStatusCode firstStatusCode,
   T response)
{
    var service = new FakeStorageService();
    var client = new StorageClientImpl(service);
    var request = requestProvider(service);

    service.ExpectRequest(request, firstStatusCode);
    service.ExpectRequest(request, new T());
    clientAction(client);
    service.Verify();
}

private static void RetryOnceHelper<T>(Func<StorageService, ClientServiceRequest<T>> requestProvider, Action<StorageClient> clientAction,
   HttpStatusCode firstStatusCode = HttpStatusCode.BadGateway)
   where T : new() => RetryOnce(requestProvider, clientAction, firstStatusCode, new T());
```

You can then call the version which accepts the response as the last parameter for `DeleteObject` where the response is just a string.